File: backend/routes/report.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from database import get_db
import crud.report as report_crud
import schemas.report as report_schema
from datetime import date
from sqlalchemy import func
from models.transaction import Transaction

router = APIRouter(
    prefix="/reports",
    tags=["reports"],
)

# 📊 查詢各分類支出的總金額與比例
@router.get("/category-summary", response_model=list[report_schema.CategorySummary])
def get_category_summary(
    user_id: int = Query(..., description="使用者 ID"),
    start_date: date = Query(..., description="起始日期"),
    end_date: date = Query(..., description="結束日期"),
    db: Session = Depends(get_db),
):
    summary = report_crud.get_category_expense_summary(
        db=db,
        user_id=user_id,
        start_date=start_date,
        end_date=end_date
    )
    return summary  # ✅ 就算 summary 是空陣列也沒關係

# 📊 查詢使用者在指定期間內的支出與收入總額
@router.get("/totals", response_model=dict)
def get_income_expense_totals(
    user_id: int = Query(...),
    start_date: date = Query(...),
    end_date: date = Query(...),
    db: Session = Depends(get_db)
):
    from models.transaction import Transaction

    income = db.query(func.sum(Transaction.amount))\
        .filter(
            Transaction.user_id == user_id,
            Transaction.type == "income",
            Transaction.transaction_date.between(start_date, end_date)
        ).scalar() or 0

    expense = db.query(func.sum(Transaction.amount))\
        .filter(
            Transaction.user_id == user_id,
            Transaction.type == "expense",
            Transaction.transaction_date.between(start_date, end_date)
        ).scalar() or 0

    return {"income": income, "expense": expense}

File: frontend/src/api/report.js
import api from './axios';

export const getCategorySummary = (userId, startDate, endDate) => {
  // 🛡 修正非法日期，例如 2025-06-31 → 2025-06-30
  const safeEndDate = fixInvalidDate(endDate);

  return api.get('/reports/category-summary', {
    params: {
      user_id: userId,
      start_date: startDate,
      end_date: safeEndDate,
    },
  });
};

export const getIncomeExpenseTotals = (user_id, start_date, end_date) =>
  api.get(`/reports/totals`, {
    params: { user_id, start_date, end_date }
  })

// 🔧 新增這個函數
function fixInvalidDate(dateStr) {
  const date = new Date(dateStr);
  if (isNaN(date.getTime())) {
    // ❌ 無法解析的日期，就強制設成今天
    return new Date().toISOString().split('T')[0];
  }
  // ✅ 輸出 yyyy-mm-dd 格式
  return date.toISOString().split('T')[0];
}
